checker/src/org/checkerframework/checker/index/upperbound/OffsetEquation.java

```diff
@@ -0,0 +1,469 @@
+            if (operator == '-' || operator == '+') {
+                index++;
+            } else {
+                operator = '+';
+            }
+
+            int endIndex = indexOf(expressionEquation, '-', '+', index);
+            String subexpression;
+            if (endIndex == -1) {
+                endIndex = expressionEquation.length();
+                subexpression = expressionEquation.substring(index);
+            } else {
+                subexpression = expressionEquation.substring(index, endIndex);
+            }
+
+            equation.addTerm(operator, subexpression);
+            index = endIndex;
+        }
+        return equation;
+    }
+
+    private static boolean isInt(String string) {
+        return string.isEmpty() || string.matches("[-+]?[0-9]+");
+    }
+
+    private static int parseInt(String intLiteral) {
+        if (intLiteral.isEmpty()) {
+            return 0;
+        }
+        return Integer.valueOf(intLiteral);
+    }
+
+    /** Returns the first index of a or b in string, or -1 if neither char is in string. */
+    private static int indexOf(String string, char a, char b, int index) {
+        int aIndex = string.indexOf(a, index);
+        int bIndex = string.indexOf(b, index);
+        if (aIndex == -1) {
+            return bIndex;
+        } else if (bIndex == -1) {
+            return aIndex;
+        } else {
+            return Math.min(aIndex, bIndex);
+        }
+    }
+
+    /**
+     * Creates an offset equation from the Node.
+     *
+     * <p>If node is an int value known at compile time, then the return equation is just the int
+     * value or if op is '-', the return equation is the negation of the int value.
+     *
+     * <p>If node is an addition or subtracted node, then this method is called recursively on the
+     * left and right hand nodes and the two equations are added/subtracted to each other depending
+     * on the value of op.
+     *
+     * <p>Otherwise the return equation is created by converting the node to a {@link
+     * FlowExpressions.Receiver} and then then added as a term to the returned equation. If op is
+     * '-' then it is a subtracted term.
+     *
+     * @param node Node from which to create offset equation
+     * @param factory AnnotationTypeFactory
+     * @param op '+' or '-'
+     * @return an offset equation from the Node
+     */
+    public static OffsetEquation createOffsetFromNode(
+            Node node, UpperBoundAnnotatedTypeFactory factory, char op) {
+        OffsetEquation eq = new OffsetEquation();
+        if (op == '-') {
+            parseRecurMinus(node, factory, eq);
+        } else {
+            parseRecurPlus(node, factory, eq);
+        }
+        return eq;
+    }
+
+    private static void parseRecurPlus(
+            Node node, UpperBoundAnnotatedTypeFactory factory, OffsetEquation eq) {
+        if (node.getTree() != null && TreeUtils.isExpressionTree(node.getTree())) {
```

Comment:
I don't understand this comment.  The smallest value of all possible values is required, so I guess the code could call the ValueATF directly, but I don't see the value in that.

The ValueLiteralNode code is duplicated by the Value Checker, so it could be removed.